### DISCUSSION TITLE
I2C gives me the Error "OSError: [Errno 5] EIO" with GY-521 (MPU6050)

### DISCUSSION THREAD

**Discussion by Moebelix** [micropython/micropython#11070] [RP2040 / Pico] [answered]
Hello everybody,

I am new to this forum, so please excuse if I am writing in the wrong area. 

I have been fighting to get a GY-521 accelerometer and gyroscope board running on a RP2040 Type C 16mb Chinese Clone, and also on a different clone.

I am able to initialize a I2C object
I am able to use the scan function and it finds the address of the MPU6050 (on the GY-521 board)
But I am not able to read anything from the MPU6050
```
from machine import Pin, PWM, I2C
from time import sleep
import utime
from imu import MPU6050
from deltat import DeltaT
import fusion

i2c = I2C(0, sda=Pin(20), scl=Pin(21), freq=400000)


devices = i2c.scan()

device_count = len(devices)

if device_count == 0:
    print('No i2c device found.')
else:
    print(device_count, 'devices found.')

for device in devices:
    print('Decimal address:', device, ", Hex address: ", hex(device), "\t\t")
        
sleep(1)

#print(i2c.readfrom(0x68, 6))

#print(i2c.readfrom_mem(0x68, 0x3B, 6))

testbuff = bytearray(6)    
i2c.readfrom_mem_into(0x68, 0x3B, testbuff,addrsize=8)
```

I get the following output:
```
>>> %Run -c $EDITOR_CONTENT
1 devices found.
Decimal address: 104 , Hex address:  0x68 		
Traceback (most recent call last):
  File "<stdin>", line 30, in <module>
OSError: [Errno 5] EIO
>>> 
```

I have tested an other GY-521 board and a PCA9685 board with the same results.

I have tested all of these boards with an arduino uno, and they worked fine (both with 3.3V and 5V)

Does anybody know what I am doing wrong?

**Comment by karfas** [comment, marked as answer]
Just search the forum for `i2c "errno 5"`. 
Most likely, your wiring misses pull-up resistors or you are using the wrong pins for i2c.

**Reply by Moebelix** [reply]
I already did this search. 

I tried setting the pull up:
`i2c = I2C(0, sda=Pin(20, pull=Pin.PULL_UP), scl=Pin(21, pull=Pin.PULL_UP), freq=400000)`

This did not help

The scan() function finds the device at the correct address, therefore I find it unlikely, that I used the wrong pins. I also tried this on pins 
1 and 0 
12 and 13

Always the same result

**Reply by Moebelix** [reply]
I am sorry, I just tried the pull ups again and it worked.
Thank you for the quick help

**Reply by robert-hh** [reply]
The soft-PullUp resistors are usually too weak (like 30kOhm). They may be good enough for a scan(), but fail at longer data. Better attach external ones in the range of 2.2K - 10k Ohm.

**Reply by Fidoooo** [reply]
Hi, could you please share the final code you used? I am having similar issue and I am getting a bit clueless here

**Reply by baldingwizard** [reply]
I had this exact problem for an AHT21 sensor. Added external 10K pull-ups and communication worked like a charm. Thanks @robert-hh - really appreciate the tip.

EDIT: Actually, I've just tried without the pullups and it's still working.... hmmmm. Perhaps it depends on noise and frequency. I only have the single device on the bus. Trying without the pullups but with reduced i2c frequency as I'd rather get away without pullups if I can (to simplify the build onto a Pico).

**Reply by baldingwizard** [reply]
I've just re-tested this without external pullups and a drastically reduced clock frequency on a Pi Pico, with the AHT21 connected directly to pins 1 (SDA), 2 (SCL), 3 (GND) and 4 (Vin) with the internal pullups enabled. The plan is to directly solder the 4-pin module to the Pico without any other external components. Note that as it's low power I'm actually powering the module using GP2 - since it's a very low-power module and well within the limits - so I can switch the module on by setting GP2 high and then read the sensor values using i2c0 on GP0 and GP1, switching it off again if I need to save power. It seems to work reliably so far (I only need a low frequency so tested with freq=4000).
